### PR TITLE
Adding keep_empty argument to list_c, list_cbind, and list_rbind. 

### LINKED ISSUE #1096: FR: New `keep_empty = FALSE` argument to `list_c()` and `list_rbind()`
I'm looking for the following behavior so that I can use `pivot_wider() %>% mutate(across(..., ~ list_c(.x, keep_empty = TRUE)))` as a more robust replacement for `tidyr::spread(convert = TRUE)` .

I can achieve it today with the lower-level `list_unchop()`, in a very clumsy way.

``` r
library(purrr)

# Existing behavior:
list_c(list("a", NULL, "b"))
#> [1] "a" "b"
vctrs::list_unchop(list("a", NULL, "b"))
#> [1] "a" "b"
vctrs::list_unchop(list("a", NULL, "b"), indices = as.list(1:3))
#> [1] "a" NA  "b"

# I'd like to write this:
list_c(list("a", NULL, "b"), keep_empty = TRUE)
#> [1] "a" NA  "b"
```

<sup>Created on 2023-08-20 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by SokolovAnatoliy**
Fixes #1096.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # purrr (development version)
 
+* Added a new keep_empty argument to list_c(), list_cbind() and list_rbind() which will keep empty elements as NA in the returned dataframe. (#1096)
```
Comment (hadley) [inline NEWS.md:3]:
```suggestion
* Added a new keep_empty argument to `list_c()`, `list_cbind()`, and `list_rbind()` which will keep empty elements as `NA` in the returned data frame. (#1096)
```

File: R/list-combine.R [hunk 2]
```diff
@@ -22,6 +22,7 @@
 #'   same size (i.e. number of rows).
 #' @param name_repair One of `"unique"`, `"universal"`, or `"check_unique"`.
 #'   See [vctrs::vec_as_names()] for the meaning of these options.
+#' @param keep_empty An optional Logical to keep empty elements of a list as NA.
```
Comment (hadley) [inline R/list-combine.R:25]: Could you try rewriting this using the formula "If `FALSE` (the default), then ...; if `TRUE`, then ...`.

File: R/list-combine.R [hunk 3]
```diff
@@ -30,17 +31,21 @@
 #'
 #' x2 <- list(
 #'   a = data.frame(x = 1:2),
-#'   b = data.frame(y = "a")
+#'   b = data.frame(y = "a"),
+#'   c = data.frame(z = NULL)
```
Comment (hadley) [inline R/list-combine.R:35]: I think this is equivalent to just `data.frame()`?

File: R/list-combine.R [hunk 4]
```diff
@@ -58,19 +63,22 @@ list_cbind <- function(
     x,
     ...,
     name_repair = c("unique", "universal", "check_unique"),
-    size = NULL
+    size = NULL,
+    keep_empty = FALSE
   ) {
   check_list_of_data_frames(x)
   check_dots_empty()
+  if(keep_empty) x <- convert_empty_element_to_NA(x)
```
Comment (hadley) [inline R/list-combine.R:71]: Are you sure this applies to `list_cbind()`? I don't know why but I think it shouldn't?

File: R/list-combine.R [hunk 5]
```diff
@@ -95,3 +103,8 @@ check_list_of_data_frames <- function(x, error_call = caller_env()) {
     call = error_call
   )
 }
+
+## used to convert empty elements into NA for list_binding functions
+convert_empty_element_to_NA = function(x) {
+  map(x, \(x) if(vctrs::vec_is_empty(x)) NA else x)
```
Comment (hadley) [inline R/list-combine.R:109]: I think this anonymous function syntax requires R 4.1? Otherwise, this implementation looks good to me!

File: tests/testthat/test-list-combine.R [hunk 6]
```diff
@@ -70,6 +70,15 @@ test_that("NULLs are ignored", {
   expect_equal(list_cbind(list(df1, NULL, df2)), vec_cbind(df1, df2))
 })
 
+test_that("NULLs are converted to NA when keep_empty = TRUE", {
+  df1 <- data.frame(x = 1)
+  df2 <- data.frame(y = 1)
+
+  expect_equal(list_c(list(1, NULL, 2), keep_empty = TRUE), c(1, NA, 2))
+  expect_equal(list_rbind(list(df1, NULL, df1), keep_empty = TRUE), vec_rbind(df1, NA, df1))
```
Comment (hadley) [inline tests/testthat/test-list-combine.R:78]: I think I'd find it a little easier to verify that this is correct if the expected value looked like `data.frame(x = c(1, NA, 1)`

**Review by hadley — COMMENTED** [review]
Thanks for working on this!

File: R/list-combine.R [hunk 7]
```diff
@@ -95,3 +105,8 @@ check_list_of_data_frames <- function(x, error_call = caller_env()) {
     call = error_call
   )
 }
+
+## used to convert empty elements into NA for list_binding functions
+convert_empty_element_to_NA = function(x) {
+  map(x, function(x) if(vctrs::vec_is_empty(x)) NA else x)
```
Comment (krlmlr) [inline R/list-combine.R:111]: Should that be something like

```suggestion
  map(x, function(x) if(is.null) NA else x)
```

?

File: R/list-combine.R (hunk 4 above)
Comment (krlmlr) [reply R/list-combine.R:73]: To me, it's unclear what type the empty column should have.

File: NEWS.md (hunk 1 above)
Comment (krlmlr) [reply NEWS.md:3]: If you like, you can add your GitHub handle after the issue: `(#1096, @SokolovAnatoliy)` .

**Review by krlmlr — COMMENTED** [review]
Thanks for taking this on!

File: R/list-combine.R (hunk 7 above)
Comment (mgirlich) [inline R/list-combine.R:111]: This might have quite a performance impact if `x` is big. A faster alternative would be something like

```r
empty <- vctrs::list_sizes(x) == 0
x[empty] <- NA
x
```

File: R/list-combine.R (hunk 4 above)
Comment (hadley) [reply R/list-combine.R:73]: It's also weird that for `list_cbind()` the name of empty columns comes from the outer name rather than the inner name.

File: R/list-combine.R [hunk 8]
```diff
@@ -58,19 +66,26 @@ list_cbind <- function(
     x,
     ...,
     name_repair = c("unique", "universal", "check_unique"),
-    size = NULL
+    size = NULL,
+    keep_empty = FALSE
```
Comment (hadley) [inline R/list-combine.R:70]: @DavisVaughan my intuition is that `keep_empty` doesn't make sense for `list_cbind()`, but I don't have anything concrete to back that up. Do you have any thoughts?

File: R/list-combine.R [hunk 9]
```diff
@@ -22,6 +22,9 @@
 #'   same size (i.e. number of rows).
 #' @param name_repair One of `"unique"`, `"universal"`, or `"check_unique"`.
 #'   See [vctrs::vec_as_names()] for the meaning of these options.
+#' @param keep_empty  An optional logical. If `FALSE` (the default), then
```
Comment (DavisVaughan) [inline R/list-combine.R:25]: I think `keep_empty` is the wrong name if `double()` isn't promoted to `NA_real_` and kept. It's more like `keep_null`.

```r
> list_c(list(1, double(), 2))
[1] 1 2
> list_c(list(1, double(), 2), keep_empty = T)
[1] 1 2
```

Comment (DavisVaughan) [reply R/list-combine.R:25]: Compare with

```r
> tidyr::unchop(tibble::tibble(x = list(1, double(), 2)), x)
# A tibble: 2 × 1
      x
  <dbl>
1     1
2     2
> tidyr::unchop(tibble::tibble(x = list(1, double(), 2)), x, keep_empty = TRUE)
# A tibble: 3 × 1
      x
  <dbl>
1     1
2    NA
3     2
```

File: R/list-combine.R [hunk 10]
```diff
@@ -30,16 +33,21 @@
...
+#' list_rbind(x2, names_to = "id", keep_empty = TRUE)
 #' list_rbind(unname(x2), names_to = "id")
-#'
 #' list_cbind(x2)
-list_c <- function(x, ..., ptype = NULL) {
+#' list_cbind(x2, keep_empty = TRUE)
+list_c <- function(x, ..., ptype = NULL, keep_empty = FALSE) {
   vec_check_list(x)
   check_dots_empty()
+  if (keep_empty) {
```
Comment (DavisVaughan) [inline R/list-combine.R:48]: We'd want to use `check_bool(keep_empty)` everywhere

File: R/list-combine.R [hunk 11]
```diff
@@ -95,3 +110,9 @@ check_list_of_data_frames <- function(x, error_call = caller_env()) {
     call = error_call
   )
 }
+
+convert_null_to_NA <- function(x) {
+  is_null <- map_lgl(x, is.null)
+  x[is_null] <- list(NA)
```
Comment (DavisVaughan) [inline R/list-combine.R:116]: `vctrs::vec_detect_missing()` would be much faster at detecting `NULL`s since we know `x` is a list.

See `tidyr:::list_replace_null()` for a robust and very fast version of this operation

File: tests/testthat/test-list-combine.R [hunk 12]
```diff
@@ -70,6 +70,24 @@ test_that("NULLs are ignored", {
...
+  )
+  expect_equal(
+    list_rbind(list(df1, NULL, df1), keep_empty = TRUE),
+    data.frame(x = c(1, NA, 1))
+  )
+  expect_equal(
+    list_cbind(list(df1, z = NULL, df2), keep_empty = TRUE),
+    data.frame(df1, z = NA, df2)
+  )
+})
```
Comment (DavisVaughan) [inline tests/testthat/test-list-combine.R:89]: If we decide to implement this then I'd like the chance to come through and write _a lot_ of tests for the 3 cases individually. The `keep_empty` logic in tidyr is quite hard to get 100% right, and required a lot of edge case tests.

**Review by DavisVaughan — COMMENTED** [review]
Dropping in some initial comments while looking at this. No need to make any changes yet though.

But also see https://github.com/tidyverse/purrr/issues/1096#issuecomment-2315466918. I think it is possible we made a small mistake by marking this as a TDD issue, since it can already be solved through `tidyr::unchop()` in a way that I feel is much more robust for the motivating example given there (i.e. the size invariants are nicer when you use `unchop()`).

**Comment by DavisVaughan** [comment]
We have decided not to implement this feature for now since:
- `tidyr::unchop()` solves the problem already in a very robust way https://github.com/tidyverse/purrr/issues/1096#issuecomment-2315466918
- We can't seem to justify the argument for `list_cbind()`, suggesting something feels off with putting it here in purrr

@SokolovAnatoliy we apologize about that! We hope you had fun at TDD anyways, and it seems like you got to do some other PRs in other repos, which is awesome! Thanks again!
